### ISSUE TITLE
ems-esp is rebooting twice a day

### ISSUE THREAD

**Issue by tp1de** [emsesp/EMS-ESP32#78]
@proddy 
Hi Paul, I recognized that my ESP32 gateway is rebooting approx. after 11-12 hours. I can't see any reason why. Free memory is not the problem. Could it be that having http get requests per device every 15 sec per api is initiating the rebooting after a while?

It's not a major issue for me. I just miss some data for approx 90 secs because of my 60 tx-delay on startup.
Could I help to resolve this problem?

**Comment by proddy** [comment]
This is going to hard to find. It's probably the web code (API) but could you try and narrow it down to get it easily reproducible. For starters
- does it happen on WiFi or Ethernet or both?
- are you sure its rebooting and not just reconnecting? (i.e. uptime sets back to 0)?
- did you see similar issues with earlier v3 releases?
- does the same thing happen if you disable the other services like MQTT, NTP, SysLog?
- are you able to see from the logs when and where it crashes? is it always the same spot?
- if you increase the poll rate from 15 seconds to less, does it crash quicker?
- if you limit the number of API calls does it help?
- would it be helpful if there was an API call to group calls to prevent too many TCP connections?

**Comment by proddy** [comment]
well spotted @MichaelDvP !

**Comment by tp1de** [comment]
@proddy @MichaelDvP 
I will test and report.

**Comment by tp1de** [comment]
@proddy @MichaelDvP 

I installed the updated version with the _fixed_ memory leak. Now it is getting worse.
I get a reboot now approx. every 13 minutes. It is a reboot of the esp32 and not just a connection lost for ems-bus.

I will try to watch the log if I can see anything ....

**Comment by proddy** [comment]
The fix @MichaelDvP did is valid. The async response buffer is only free'd after a send (https://github.com/me-no-dev/ESPAsyncWebServer#request-life-cycle).

**Comment by tp1de** [comment]
Now it is running since > 30 minutes. But I recognized the following in respect to web-ui system log:

- When loglevel is set to error I get 1-2 crc errors shown per minute
- When loglevel is set to debug then no crc errors are shown - that's strange -or ?

I switched off the web-ui system log and will wait if this has an influence.

**Comment by proddy** [comment]
> When loglevel is set to error I get 1-2 crc errors shown per minute

That is really bad. What does `show ems` tell you from the console?

![Screenshot 2021-06-29 130333](https://user-images.githubusercontent.com/1230712/123786745-55453080-d8da-11eb-9f15-0ef0867543b4.png)

**Comment by tp1de** [comment]
What do you mean with bad? Too many crc errors?
**But why aren't the errors shown in debug mode?**
Actually the esp is running since 1:20 hours and system log-level is off:

![grafik](https://user-images.githubusercontent.com/36480251/123787734-7a866e80-d8db-11eb-9358-ac1e4b1903f0.png)

**Comment by proddy** [comment]
> What do you mean with bad? Too many crc errors?
But why aren't the errors shown in debug mode?

1-2 CRC errors every minute is not good. Are you sure you are seeing this?
Errors are also shown when the log level is DEBUG. If you're not seeing any probably there aren't any! It's hard to tell without real logs, screenshots etc

**Comment by MichaelDvP** [comment]
Reboots and crc-errors can be induced by a weak/unstable powersupply, thin cables, etc. Last in #72, and many times before.

**Comment by tp1de** [comment]
> Reboots and crc-errors can be induced by a weak/unstable powersupply, thin cables, etc. Last in #72, and many times before.

Yes I am certainly aware, but I haven't changed anything on wirering and/or power supply. 
I haven't had this "instability" in older versions for some months. 
I will test in former builds if I will have the same behaviour.
0.9% crc errors on rx should not be a major problem as far as I can judge.

**Comment by tp1de** [comment]
The gateway rebooted again after approx 3,5 hours. I had log level on "off". After a reboot it is reset to "info".

![grafik](https://user-images.githubusercontent.com/36480251/123807717-6600a100-d8f0-11eb-87b5-75cefb54d996.png)

**Comment by proddy** [comment]
I would ignore the CRC errors and how they are reported. It's normal to see CRC errors. Logs are only shown when the browser window is open and it's always set to INFO. So all normal there.

To find out why it's rebooting, try playing with the API calls (lower the number of calls, turn it off) until you narrow down what is causing it

**Comment by MichaelDvP** [comment]
The log looks like bus-Id is set to 0x0A and all crc-errors are on the master-echo. Strangely the stat above shows no tx-fails, so the other devices reply to emsesp, only emsesp itself is disturbed while sending. Maybe bus-connection is not good. 

I'm trying to reproduce:
I have activated the tp1de-ioBroker-adapter in actual version running now for seven and a half hours without reboot (after the fix). Also mqtt is active with my single-value extension and qos 1, both giving a lot of more traffic. 
Jack-powered Gateway S32 prototype, Jack-cable 0,5m from BBQKees.
Rx: 35159, crc-fails 1
Tx: 3915, fails 0
Dallas-reads:  18701, fails 0
Mqtt publishes 14873, fails 0
Devices: Boiler, Controller, Mixer, Solar, Thermostat, 7 Dallassensors powered from 3.3V, 1 analog sensor powered from 5V.
Syslog info enabled, AP disabled, NTP enabled, OTA disabled
Heap 191k, free block 107k (unchanged in this 7,5 hours)

All looks stable to me, i don't think i will get a reboot in next hours, but we'll see tomorrow.

**Comment by MichaelDvP** [comment]
Uptime: 001+01:35:15
Rx: 14089, 17 fails
Tx: 23549, 0 fails
mqtt 33015, 0 fails
heap 191k, 107k free block

I can not reproduce the reboots. Software (mqtt/api/ioBroker-adapter)  seems stable, i think it's somewhere in hardware/connection/power.

**Comment by proddy** [comment]
there's a guy on Discord that's also getting reboots. Perhaps we should re-introduce the code that kept the reboot reason and part of the stack to help identify which is causing it. If it's a Watch Dog reset then I would be worried

**Comment by tp1de** [comment]
I updated again with the fixed version from @MichaelDvP and made a hard reboot.
I can confirm that it seems stable now for more then 24 hours. 

The higher crc error rate should not be the cause for reboots.
My 0.8% crc error rate is related to connection the ems-esp gateway to the ems-bus outbound of mm100 mixer module (2 m twisted pair cable) to keep service jack free for maintenance activities. That's the reason why I have selected a different bus id to allow the installer on yearly maintenance to connect by service jack and laptop.

But I believe the suggestion of @proddy is good to keep track of reboot reasons.
BTW yesterday I could initiate reboots by webui from my mobile phone (android 10 / chrome browser) by requesting a couple of times a request to refresh system status screen. Today it is working.

**Comment by tp1de** [comment]
Just a small observation: freemem info by api is always lower then by mqtt heartbeat (174 vs. 180).

**Comment by proddy** [comment]
> Just a small observation: freemem info by api is always lower then by mqtt heartbeat (174 vs. 180).

https://github.com/emsesp/EMS-ESP/issues/569

**Comment by MichaelDvP** [comment]
> (2 m twisted pair cable)

how thin? Try with 0,75mm^2. The crc-errors only while sending indicating a voltage drop over the line.

I see that your KM200 produces a lot of traffic, i have ~75 messages per minute, you receive 175 per minute. That is something i can not reproduce in my system.

**Comment by tp1de** [comment]
> how thin? Try with 0,75mm^2. The crc-errors only while sending indicating a voltage drop over the line.

Yes I am aware but this only tempory. (.... since 5 months ... ) I will install the ems-esp close to the mm100 and then cable will then be 10 cm.

> I see that your KM200 produces a lot of traffic, i have ~75 messages per minute, you receive 175 per minute. That is something i can not reproduce in my system.

What kind of data for trafic are you refering to Michael? ems telegrams?

**Comment by MichaelDvP** [comment]
> What kind of data for trafic are you refering to Michael? ems telegrams?

Yes, traffic on the ems-bus. But i don't think that the higher telegram-rate can cause resets after a couple of hours. The number of polls on the ems-bus is higher and does not overload the emsesp. I only mentioned it, because it is a difference between your system and mine.

**Comment by tp1de** [comment]
> i have ~75 messages per minute, you receive 175 per minute

Just for interest. Where do you get the figures from and what do you mean with KM200?
KM200 gateway with device id 0x48 or the km200 polling out of the ems-esp ioBroker adapter?
Where do you identify 175 messages per minute on my system?

**Comment by MichaelDvP** [comment]
You wrote yesterday:
>> Actually the esp is running since 1:20 hours and system log-level is off:

and the screenshot shows 14212 received messages, giving 14212/80 = 177 messages per minute.

Because our systems are not very different (me: GB125-MC10, RC35, MM10, SM10, you GB125-MC110, RC300, MM100) i think the km200 makes the difference, reading active the devices.

**Comment by tp1de** [comment]
OK understood. Actually longterm it is rather 193 messages per minute.

I believe this is the result of polling km200 every 90 seconds within ioBroker Adapter  for 119 fields w/o recordings.
So approx 80 per minute + km200 is polling all recordable fields every minute to create recording statistics (power + temp).
These are another 20 messages.

So km200 is increasing message traffic .

**No reboots on ems gateway so far anymore!**

**Comment by tp1de** [comment]
I got an other _reboot_ at 3 am this night. Or could it be that something else is resetting uptime?

**Comment by proddy** [comment]
when EMS-ESP boots it will send out an MQTT topic called `info` with a payload like
```json
{"event":"start","version":"3.1.2b0","ip":"192.168.1.134"}
```
This is a good indication on whether the device has physically rebooted itself.

**Comment by MichaelDvP** [comment]
@tp1de 
Have you tested a different power supply or jack-power? 
Most of the other users don't report reboots, it's pobably not the software, but something in your system.

**Comment by tp1de** [comment]
> 
> 
> when EMS-ESP boots it will send out an MQTT topic called `info` with a payload like
> 
> ```json
> {"event":"start","version":"3.1.2b0","ip":"192.168.1.134"}
> ```
> 
> This is a good indication on whether the device has physically rebooted itself.

I will watch this. Since I have restarted manually afterwards I see this mqtt entry from the manual start. I was wondering that my ioBroker Adapter with 15 sec polling have not recognized any errors. I will recheck my code to report http get errors during reboot..

But I believe it would be usefull to have the last restart status and time beeing reported in mqtt and in api v3 in http://ems-esp/api/system unless it is just the uptime which is reset when rebooted / restarted..

**Comment by tp1de** [comment]
> 
> 
> @tp1de
> Have you tested a different power supply or jack-power?
> Most of the other users don't report reboots, it's pobably not the software, but something in your system.

I will do starting today with the following steps:

1. Connect with thicker and shorter wires to mm100 ems-bus out and watch crc errors
2. Connect to jack but keep it usb powered
3. connect to jack and jack-powered

I will report crc errors effect and will watch if this has an impact on rebooting. This might take a couple of days.
Nevertheless the rebooting does not effect long-term stability of my ioBroker adapter I think we should find the root cause.

**Comment by tp1de** [comment]
I will report my crc results in #80 
In the meantime I have updated my ioBroker adapter to show an error message when no devices found during reboot.

**Comment by tp1de** [comment]
#80 is understood and no issue for me anymore. So let me come back to the rebooting issue.
I just had a reboot after approx. 10 hours. ems-esp now connected with 10 cm wiring to mm100 and usb powered. I will change usb power adapter but I do not believe that this is the reason for reboots.

What else can I do / test?

**Comment by tp1de** [comment]
I had mqtt explorer running during reboot.
The only thing I recognized was a small negative spike on freemem shortly before rebooting:
![grafik](https://user-images.githubusercontent.com/36480251/124245122-f53fcb80-db1f-11eb-9e9d-1ce5d177889d.png)

**Comment by proddy** [comment]
We could add the uptime_sec to `/api/system/info`. It's just one line to add.

**Comment by proddy** [comment]
> I just had a reboot after approx. 16 hours. ems-esp now connected with 10 cm wiring to mm100 and usb powered. I will change usb power adapter but I do not believe that this is the reason for reboots.

I was looking into how to capture the reboot reason and backstack trace on an ESP32 and it's not easy, unlike the ESP8266.

**Comment by tp1de** [comment]
> 
> 
> We could add the uptime_sec to `/api/system/info`. It's just one line to add.

This would be good. Any idea how to find the reason for rebooting?

**Comment by MichaelDvP** [comment]
> I was looking into how to capture the reboot reason and backstack trace on an ESP32 and it's not easy, unlike the ESP8266.

For the reason there is an example [here](https://github.com/espressif/arduino-esp32/blob/master/libraries/ESP32/examples/ResetReason/ResetReason.ino).

**Comment by proddy** [comment]
> This would be good. Any idea how to find the reason for rebooting?

No idea. As I mentioned in the earlier comment https://github.com/emsesp/EMS-ESP32/issues/78#issuecomment-870266749 the way to find it is by process of elimination, i.e. simply stopping services and slowing down API calls to see if it makes a difference.  I suspect its probably the # API calls causing the AsyncWebServer to WDT. There are many reported issues in that's projects GitHub repo. But...we have around 1700 EMS-ESP users out there and no one else has reported reboots like this

**Comment by tp1de** [comment]
>  I suspect its probably the # API calls causing the AsyncWebServer to WDT.

@MichaelDvP Since you have installed the ioBroker adapter as well. Do you have rebooting issues?

Since my ESP32 is powered by USB and yours as far as I have understood by service jack, this is the only key difference - isn't it? 
What else could bring the ESP32 to reboot? Could it depend on different BBQKees board version?

**Comment by proddy** [comment]
@tp1de if you turn off the API polling, does it still crash? If so we can experiment with lowering the Tx power to help on voltage fluacations and I can create a small stress test. For that test I would need to know which API calls are you sending and the frequency,

**Comment by MichaelDvP** [comment]
> @MichaelDvP Since you have installed the ioBroker adapter as well. Do you have rebooting issues?

No. As written [here](https://github.com/emsesp/EMS-ESP32/issues/78#issuecomment-871230550)
>  Could it depend on different BBQKees board version?

Thanks to BBQKees he has spend some boards to me. I've tested gateway v1.5 with MH-ET32S, v1.7 with a different MH-ET, Gateway S32 prototype and S32 series. Only eth connection is untested with my system. No reboots.

**Comment by tp1de** [comment]
> if you turn off the API polling, does it still crash?

I really do not want to do this, because I will loose the historic data in ioBroker. But I could change pooling frequency.

>  For that test I would need to know which API calls are you sending and the frequency,
The ioBroker adapter has to phases. First all single fields are read per http get but only once on start.
Then there is a fixed polling every 15 seconds with
/ems-esp/api/system --> returns json with info, status and devices (in my case 3)
/ems-esp/api/thermostat
/ems-esp/api/boiler
/ems-esp/api/mixer

So in total 4 API get requests per 15 seconds. Total loop duration with http gets and writing the states into ioBroker (Redis) is approx 1 sec.

**Comment by MichaelDvP** [comment]
I think the API polling frequency is given by the ioBroker-Adapter and should be the same for me. 
Have you mqtt enabled? What sending frequency?
If it is a power issue, i mostly compile with `board_build.f_cpu = 160000000L ` in `pio_local.ini`, saving a bit power and is fast enough. (afair current for esp32: 80MHz-40mA, 160Mhz-50mA, 240MHz-70mA)

**Comment by tp1de** [comment]
Since @MichaelDvP does not have any rebooting issue with the same polling frequency API polling should not be the reason.
MQTT is enabled with 10 seconds intervals.

**Comment by tp1de** [comment]
> If it is a power issue, i mostly compile with `board_build.f_cpu = 160000000L ` in `pio_local.ini`, saving a bit power and is fast enough. (afair current for esp32: 80MHz-40mA, 160Mhz-50mA, 240MHz-70mA)

I do not compile by myself. I use the ready to use builds. https://github.com/emsesp/EMS-ESP32/releases/tag/latest

I will test now further on with the following steps:

1. I already changed USP power adapter - let's see if this has an influence
2. I will switch off MQTT
3. I will increase API polling to 60 sec

**Comment by MichaelDvP** [comment]
For what do you need api and mqtt in this high frequency? 
Wifi-sending is power-hungry. Switching off mqtt when using api is a good idea, 
I don't see any usecase for this, except for testing purpose.

**Comment by tp1de** [comment]
> For what do you need api and mqtt in this high frequency?
> Wifi-sending is power-hungry. Switching off mqtt when using api is a good idea,
> I don't see any usecase for this, except for testing purpose.

@MichaelDvP Yes you are right. I basically do not need MQTT anymore. 
Wasn't 10 seconds set as standard when switching on MQTT?

I started with MQTT and then switched to API V3 to get the single field infos (enum /units etc) and some reconnection problems I had with MQTT. I still have some MQTT-Parts in Nodered - but I can switch them off.

The 15 secs polling in within the Adapter is a result to review the modulation control mechanismen of my boiler.
I had quite some discussion with Bosch, since my Boiler always starts with 52% modulation for approx 2 minutes and mixer control is extremly slow. So boiler starts and mixer is not open for 2-4 minutes. As a result boiler is takting much more then necessary. To have "realtime" data I used the graphs from ioBroker and sent them to Bosch/Buderus. No solution from them yet.
I will try to influence the mechanismen by myself within the ioBroker adapter to reduce takting. Testing is postponed anyhow until heating period starts again .... 

But let's see what happens if I go for (2) and (3). I will report.

**Comment by proddy** [comment]
something to try later is compiling with the `-DEMSESP_WIFI_TWEAK` flag which lowers the WiFi power & strength, reducing power compensation of the circuit

**Comment by proddy** [comment]
and another thing @tp1de if you connect to a laptop via serial, when it crashes you can capture the trace which is printed to the serial console. https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/fatal-errors.html

**Comment by tp1de** [comment]
I started with new USB power supply and reducing MQTT frquency to 60 seconds. I will report the outcome.

The rebooting itself is not a severe issue for me since it is so fast that the ioBroker adapter only recognize the tx delay in respect to the km200 adapter (now 30 seconds).

**Comment by proddy** [comment]
> I had quite some discussion with Bosch

@tp1de I'm sure you're aware but please try and keep EMS-ESP low profile when you talk to Bosch. Although what we're doing is not illegal, it is a bit naughty and I don't want to get shut down by any copyright infringement

**Comment by tp1de** [comment]
> 
> 
> > I had quite some discussion with Bosch
> 
> @tp1de I'm sure you're aware but please try and keep EMS-ESP low profile when you talk to Bosch. Although what we're doing is not illegal, it is a bit naughty and I don't want to get shut down by any copyright infringement

Don't worry, I had this data already from KM200 and DS18B20 sensors where I polled 10 values every 15 secs. 
EMS-ESP was not mentioned at all.

I just tried to understand and discuss if Bosch or an Installer could adjust PID control parameters on my Boiler/Master Controller. 
Even it is more then 40 years ago when I studied electrical engineering / control systems, I understand that the control parameters are not ideal and my boiler is too large for my two houses, even I bought the boiler with 25% lower nominal power  then calculated by an energy consultant. But so far no good response from Bosch and my installer has no glue!

BTW: I got the hint of thermal problems within ESP32 while being in the BBQKees housing. 
I asked in the ioBroker Forum on the thread of my adapter for rebooting experience and got one reply so far (stable w.o. reboots).
I have seen that the ESP32 has an internal temp sensor which could be read and data sent by API / MQTT.

What do you think: Could it be thermal problems (overheating) as well which causes the reboots?
Or could it me my board - _AZDelivery ESP32 D1 Mini NodeMCU_ - which is not stable?

**Comment by tp1de** [comment]
I changed MQTT frequency to 60 secs but I had 3 reboots this night.  Now I switched MQTT off.

@proddy I have seen that you updated firmware and added uptime_sec. Thanks.
BUT you changed names als well and removed the "#" - exception are #dallas ....
Within my ioBroker adapter I take this names 1:1 for creating ioBroker states and SQL-history - now I have them twice and history is on old names. Please do not change too often, since this needs manual intervention.
I would advice to change dallas info as well and remove the #.

**Comment by tp1de** [comment]
@proddy  Any chance to read internal temperature sensor and publish values?
https://circuits4you.com/2019/01/01/esp32-internal-temperature-sensor-example/

**Comment by MichaelDvP** [comment]
Internal temperature sensor was only in first silicon, later esp32 do not have a sensor. Check the [old datasheet here](https://cdn.sparkfun.com/datasheets/IoT/esp32_datasheet_en.pdf)  and [actual datasheet here](https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf). The internal sensor was not calibrated and only suitible for measuring increase/decrease, not absolut temperatures, as described in the old datasheet. Better you connect a dallas and put it into the case.

BTW: my system is running now for 14h without reboots. IoBroker-Api enabled and mqtt with 5 sec for all.

**Comment by tp1de** [comment]
@MichaelDvP thanks for feedback. Are you using BBQKees housing? Which ESP32 board?

Somehow I feel that it might be a problem with my board. I will open the housing for better air ventilation and see if this makes a difference.

**Comment by MichaelDvP** [comment]
> Are you using BBQKees housing? Which ESP32 board?

Actually BBQKees S32 Gateway prototype in closed original housing, But i switch sometimes to Gateway v1.5 with MH-ET D1 mini32, also in the closed housing with printed side-panel. But i don't think it's temperature, the esp32 modules are specified for -40 - 85 °C environment at full power, If you don't mount it below the boiler-insulation, these temperatures are never reached.

**Comment by tp1de** [comment]
> Actually BBQKees S32 Gateway prototype in closed original housing, But i switch sometimes to Gateway v1.5 with MH-ET D1 mini32, also in the closed housing with printed side-panel. But i don't think it's temperature, the esp32 modules are specified for -40 - 85 °C environment at full power, If you don't mount it below the boiler-insulation, these temperatures are never reached.

Since the ESP32 was handwarm while opening the housing, I think you are right.
I started with gateway board v1.6 and used it for some months and since some weeks I am using v1.7 with the updated ems screw terminal.

Rebooting could be caused by my ESP32 - AZDelivery ESP32 D1 Mini NodeMCU - which is not stable or maybe the v1.7 board.
Since I do not have a second ESP32 I will test the v1.6 gateway board in the meantime.

**Comment by tp1de** [comment]
v1.6 gateway board rebooted after approx. 9 1/2 hours. Last try is with jack-powered v1.6 gw-board.

**Comment by tp1de** [comment]
Before going to bed late I recognized that EMS-ESP lost the network connection (w.o. reboot)

![grafik](https://user-images.githubusercontent.com/36480251/124368966-0fd08c80-dc67-11eb-997c-21554f2ce02f.png)

**Comment by proddy** [comment]
As I keep saying turn off NTP, MQTT and any other unnecessary services. If WiFi is reconnecting it smells like a power issue.

**Comment by MichaelDvP** [comment]
I'm stopping my test with high datarate now, It's 37 hours without reboot, 55000 api calls, 203000 mqtt publishes, 163000 telegrams received (5 crc fails), 41000 telegrams sent, 95000 dallas read. It is not the software, not api, not mqtt that causes the reboots. 
@tp1de you should check with stable power and if this does not help, change the esp32 module. Maybe it has a ESD-damage or bad soldering.

**Comment by tp1de** [comment]
> As I keep saying turn off NTP, MQTT and any other unnecessary services. If WiFi is reconnecting it smells like a power issue.

I had everything off and even reduced the API polling rate to 60 secs and still had reboots. This time when connected by service jack I do not see reboots anymore but I had twice this night WiFi connection losses. I do not think it's NTP or MQTT services.
I do have another idea, what could make the connection losses:

I do have a Mesh WiFi Network consisting of AVM Fritzboxes and Wlan Repeaters. A FB 6591 is connected to Cable-Internet in my Office 1st floor and is the Mesh Master. A second Fritzbox 7590 is conected by LAN cable and located in the cellar and acting as Mesh Repeater. Another AVM repeater is within Mesh and connected by Wifi. 

I had already in the past connection problems to Mesh with EMS-ESP (connecting to poor signal strength repeater) but other users seems to have an issue with Mesh as well - see #762. 

My WiFi networks autoselects channels and is reducing bandwith dynamically when recognizing too many other WiFi networks on same channel. I do have 21 WiFi networks arround my house. In the log of my main router I can find an entry, that shortly before EMS-ESP Wifi connection was lost, this happened and Wlan bandwith was reduced from 40 MHz (EMS-ESP: n/Wi-Fi 4, 40 MHz, WPA2, 1 x 1) to 20 MHz. Then connection was lost and it took 3 minutes to reconnect. (no reboot so far). 
Other WiFi connected devices had no issue. I just see that one of my raspberry pi's connected to same accesspoint always uses 20 MHz bandwith and not 40 like EMS-ESP.

I now switched off this dynamic bandwith adjustment feature and will see what happens. Uptime is now 13.5 hours.

BTW: It would be good to see local time rather then uptime in system-log.

**Comment by tp1de** [comment]
> I now switched off this dynamic bandwith adjustment feature

and had after 6:20 hours another WiFi connection lost. Reconnection took 16 seconds. Since I connected to service jack I only see the WiFi reconnection reporting in system log and my ioBroker adapter misses 1 http get, but I get no rebooting anymore since 17:30 hours. 

I do have more then 10 Wlan-devices connected to this accesspoint. As far as I understand the document https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi.html#esp32-wi-fi-configuration recommends in this case to go for HT20 as standard:

_ESP32 supports Wi-Fi bandwidth HT20 or HT40, it doesn’t support HT20/40 coexist. esp_wifi_set_bandwidth() can be used to change the default bandwidth of station or AP. The default bandwidth for ESP32 station and AP is HT40._

_In station mode, the actual bandwidth is firstly negotiated during the Wi-Fi connection. It is HT40 only if both the station and the connected AP support HT40, otherwise it’s HT20. If the bandwidth of connected AP is changes, the actual bandwidth is negotiated again without Wi-Fi disconnecting._


_Theoretically the HT40 can gain better throughput because the maximum raw physicial (PHY) data rate for HT40 is 150Mbps while it’s 72Mbps for HT20. However, if the device is used in some special environment, e.g. there are too many other Wi-Fi devices around the ESP32 device, the performance of HT40 may be degraded. So if the applications need to support same or similar scenarios, it’s recommended that the bandwidth is always configured to HT20._

Any chance to test this?

**Comment by proddy** [comment]
Can you build the firmware yourself? If so add -DEMSESP_WIFI_TWEAK to the `pio_local.ini` and I'll make sure HT20 is inforced using the call `esp_wifi_set_bandwidth(ESP_IF_WIFI_STA, WIFI_BW_HT20);`. Also I noticed you have a RPi in your network. It comes with a syslog service so you could get EMS-ESP to dump all the logs there, which will give you more insight into what is happening instead of using the console or Web System Log.

**Comment by tp1de** [comment]
I have't compiled by myself since a long time. I dowloaded source from last build. Is this complete?
I just tried compiling but I am getting errors;
![grafik](https://user-images.githubusercontent.com/36480251/124389259-a1cbaa00-dce6-11eb-97a9-ab1fe4ccc1fc.png)

BTW: I just got a reboot even when switched of MQTT, NTP and being connected to servicejack-powering. I am getting frustrated.

Yes I have different syslog services up and running. Which loglevel shall I use to avoid too much entries? Is "error" good enough?

**Comment by FredericMa** [comment]
I'm having the same issue. Mine restarts at random intervals.
For me, the issue started from version 3.1.0 but I didn't had the time to look at it.
I thought that it maybe was a memory issue so I cleared the flash memory and reflashed EMS-ESP today but it didn't solve the issue.
I reinstalled it around 7PM (yellow mark) on the chart) and as you can see it restarted 15 minutes ago.

Here is a screenshot of the uptime chart:
![image](https://user-images.githubusercontent.com/7840263/124397896-10bef800-dd13-11eb-8bbe-6681ee1739f2.png)

**Comment by tp1de** [comment]
I got another reboot too. Syslog Server has no further information other than the restart info.

**Comment by proddy** [comment]
> I have't compiled by myself since a long time. I downloaded source from last build. Is this complete?

@tp1de @FredericMa I've added a new PlatformIO target called 'debug' which compiles with debug strings and uses HT20 WiFi. Easiest way is to use Visual Studio Code with Pio and select Project Tasks->debug->Upload and Monitor with EMS-ESP connected to USB. When EMS-ESP crashes it will show the decoded stack dump and point to where line and which object file caused it. Best to turn off MQTT and NTP to save on chatter.

**Comment by tp1de** [comment]
@proddy Thanks I will wait until tomorrow. I need to install than VSC on my Laptop. I use my Desktop for SW-development.

**In my case I am quite confident that I might have found the reason for reboots.** 
Extending the WLAN log in my Fritzbox Router I recognized heavy activity within my Mesh-Network of changing WiFi-Bandwidth and/or changing channels due to heavy workload of other network-devices. This happens 1-2 times an hour until I switched this off yesterday. (It is on by standard setup since years and I haven't recognized yet).

Actually everthing is stable and I do not have reboots anymore since 17 hours - but I will wait at least another 24 hours.

Could it be that API V3 does not recognize properly if WiFI is temporarily down?
Reconnection works within 1-2 seconds. This does not seem to be any problem.

> I've added a new PlatformIO target called 'debug' which compiles with debug strings and uses HT20 WiFi.

What is a PlatformIO target? I had problems compiling your code since WWWData.h was missing. 
Is the code correct now and do I need to compile?

**Comment by proddy** [comment]
> Could it be that API V3 does not recognize properly if WiFI is temporarily down?

I tested, turned off WiFi and EMS-ESP goes into Access Point mode (if configured to do so) and then tries to re-connect. It works and it doesn't crash.

**Comment by tp1de** [comment]
@proddy 
>  tested, turned off WiFi and EMS-ESP goes into Access Point mode (if configured to do so) and then tries to re-connect. It works and it doesn't crash.

I got another reboot after approx. 20 hours uptime. 
In the router-log I can see that WiFi was disconnected and reconnected within less then a second.
So reconnection of WiFi works but nevertheless there seems to be a whatchdog reset which keeps EMS-ESP to reboot.
(Hardware or Software reset I can't judge. But I had similar crashes in my ioBroker adapter when doing async http get/post requests without proper try/catch error handling).

**Comment by tp1de** [comment]
> What is a PlatformIO target? I had problems compiling your code since WWWData.h was missing.
> Is the code correct now and do I need to compile?

@proddy I have installed VSC and platformio plugin on my laptop. So I am ready to do more tests. 
Actually v3.1.2b2 with IPV6 is up since 11 hours. 

I need a link to software code for compilation. 
Actually there are references within the code to an include directory and WWWData.h which gives errors on compilation.

And please explain a bit more detailled what is a "platformio target" and which steps I have to to for debugging.
I am just used to download code into a directory, open this directory in vsc then going to pio tasks and making an ESP32 build which I then upload to EMS-ESP by web-ui.

**Comment by FredericMa** [comment]
@proddy I'll try to test it this evening!

**Comment by tp1de** [comment]
@FredericMa 
> I'll try to test it this evening!

So you know how to do it and you can compile?

**Comment by MichaelDvP** [comment]
@tp1de 
WWWData.h is generated by the prebuild script and contains the web-data. You need to have NPM installed. If you run the build process is it generated automatically. (to build manual type: cd interface,  npm install, npm run build) 

To use the debug-task you have to rename `pio_local.ini_example` to `pio_local.ini` or copy the [env:debug] chapter to your pio_local.ini.
Then you can select `platformio`-icon on the left side, choose `debug`-task and select `upload and monitor`.
 
![pio](https://user-images.githubusercontent.com/59284019/124592805-27be3100-de5e-11eb-93e1-1f40a1ea5b8a.png)

**Comment by proddy** [comment]
thanks Michael. And F1->Git: Clone->`https://github.com/emsesp/EMS-ESP32.git` 

and bottom left (where it shows dev2 in Michael's setup) choose the `origin/dev` branch.

**Comment by tp1de** [comment]
@proddy @MichaelDvP 
I got errors with npm:
![grafik](https://user-images.githubusercontent.com/36480251/124595224-0ad72d00-de61-11eb-93e1-c9617c8be5dc.png)

Do you have an idea what to do?

**Comment by tp1de** [comment]
Since I develop for ioBroker I have installed: node v14.17.3 and npm v6.14.13

[log.txt](https://github.com/emsesp/EMS-ESP32/files/6769844/log.txt)

**Comment by proddy** [comment]
ok! so far so good!

**Comment by tp1de** [comment]
I just need to understand how to select the right usb port for esptool. Automatically the wrong one is choosen.
And which screen to watch afterwards Terminal or Debug Console?

**Comment by proddy** [comment]
> I just need to understand how to select the right usb port for esptool. Automatically the wrong one is choosen.
And which screen to watch afterwards Terminal or Debug Console?

in the PlatformIO menu (icon on left bar) go to PIO Home->Devices and it'll show which COM ports the USB is using. Then add this to your pio_local.ini like `upload_port = COM3`. After "Upload and Monitor" it will open a new window with the EMS-ESP console. To test it, in EMS-ESP, type "test crash" and it will force EMS-ESP to crash. Hopefully you'll see the stack dump on the screen and also the line where it crashed (test.cpp).

**Comment by tp1de** [comment]
> in the PlatformIO menu (icon on left bar) go to PIO Home->Devices and it'll show which COM ports the USB is using. Then add this to your pio_local.ini like `upload_port = COM3`

This worked for upload. Then for monitor I have to select the port manually again. So far so good.

> After "Upload and Monitor" it will open a new window with the EMS-ESP console.

This is not happening. On terminal a lot of output so fast that I can't read anything- all telegrams and the EMS-ESP is not connected to Wifi. It seems that the EMS-ESP is permanently rebooting..... I see some register dump but so fast that I can read it on the screen.

**Comment by proddy** [comment]
> This is not happening. On terminal a lot of output so fast that I can't read anything- all telegrams and the EMS-ESP is not connected to Wifi. It seems that the EMS-ESP is permanently rebooting..... I see some register dump but so fast that I can not read it on the screen

try again without the 'debug' target. Pick the 'esp32' one and see if that works. It uses a different partition.

**Comment by tp1de** [comment]
Since I had no time yesterday anymore I compiled without debug and uploaded and monitored by COM7.
EMS-ESP rebooted in the meantime - on the VSC-console you see (before reboot):

![grafik](https://user-images.githubusercontent.com/36480251/124743266-dbd2c100-df1d-11eb-8fd7-31134949e1ce.png)

after new start:

![grafik](https://user-images.githubusercontent.com/36480251/124743608-2eac7880-df1e-11eb-9a48-1c68024a1575.png)

**Comment by proddy** [comment]
weird. check that the monitor speed is 115200 in your platformio.ini?

**Comment by tp1de** [comment]
I got this screen after some hours and before reboot.
Now with b3 update that's the actual picture.
![grafik](https://user-images.githubusercontent.com/36480251/124755454-adf47900-df2b-11eb-9c29-a0cedd0c206c.png)

**Comment by tp1de** [comment]
Just rebooted again after 15 minutes with an empty screen as above .....

**Comment by tp1de** [comment]
Finally I have a dump for the rebooting:

![grafik](https://user-images.githubusercontent.com/36480251/124998052-015de880-e04c-11eb-8c11-d05eda80a7e9.png)

**Comment by tp1de** [comment]
@proddy @MichaelDvP 

Since I have the ESP8266 running in parallel to the ESP32 it's just the ESP32 which crashes after a couple of hours.
Do you have any idea what could be the root cause?

**Comment by tp1de** [comment]
A new dump (2 screens since longer)
![grafik](https://user-images.githubusercontent.com/36480251/125051504-4e20de00-e0a3-11eb-946f-ff598f151b76.png)
![grafik](https://user-images.githubusercontent.com/36480251/125051705-86282100-e0a3-11eb-991f-d6f6296af6d7.png)

**Comment by MichaelDvP** [comment]
Hm, both crashes are [`LoadProhibited`](https://docs.espressif.com/projects/esp-idf/en/v4.2/esp32/api-guides/fatal-errors.html#loadprohibited-storeprohibited) at PC 0x400014FD, trying to load something from garbage address 0x60c20 with `vfprintf.c`. 
I suspect a log message with no format specifier containing a `%` and misinterpreted as format. Maybe in 'pretty_telegram()`.

@tp1de have you tried with all logs off?

**Comment by proddy** [comment]
yes it looks like a buffer overflow trying to reference either an empty telegram or a filling a telegram buffer and exceeding the allocated space (32 bytes max, 27 for the message length). We could certainly add more checks in the code to ensure there is no overflow. It would be interesting to see if the crashes still occur with all services turned off (syslog, mqtt, ntp, api) and running only one EMS-ESP on the ems bus. If that still crashes then turn the automatic Tx off.

**Comment by tp1de** [comment]
> @tp1de have you tried with all logs off?

Yes I tried before all services off and then switched syslog on again as recommended by Paul.
I switched off everything now again except of the API V3.

The second gateway I just connected 3 days ago to test the ioBroker Adapter with V2 too - it seems to work now as well.
The reboots I have since a long time like versions > 3.1.

I will be out a week from Sunday onwards, but I will watch remote just without Laptop console.
What is the difference in sence of telegram handlings compared to old v2 which is stable?
Or might it be a V3 API topic?

**Comment by proddy** [comment]
Not entirely related, but I was reading on https://tasmota.github.io/docs/Energy-Saving/ that perhaps putting back the delay(1) we had in emsesp::loop() would save some power consumption

**Comment by MichaelDvP** [comment]
I think i've finally found the cause, a typo that was not detected by code-checker and compiler (since 23.4.2021).

**Comment by tp1de** [comment]
I loaded the new firmware and will see if it is stable now. Rebooting was between 1 and 24 hours.
I will report tomorrow.

**Comment by FredericMa** [comment]
I've also uploaded the new firmware. TX-fails is 1 now while it remained 0 before so this looks good. 1h40min uptime and counting.

**Comment by MichaelDvP** [comment]
@FredericMa, afair you have a Junkers thermostat. As mentioned in #77 i have changes the junkers-mode published values to match the command values (also to make the selection-box in web possible). Can you check if this is all ok?

**Comment by FredericMa** [comment]
Uptime is now 24h without a reboot. I would consider this as solved since I had at least 5 reboots within 24h before. Thanks!

**Comment by tp1de** [comment]
@MichaelDvP my uptime is now 32 hours. 
I think that you have found the error and corrected it.
I would recommend to wait for another 48 hours before finally closing this issue

**Comment by tp1de** [comment]
After 3 days without reboot I believe the root cause was solved by @MichaelDvP.
Thanks Michael for your support.

**Comment by proddy** [comment]
yes amazing Michael that you found the error. And consistent with Thomas' first stack dump with `retry_tx()`:

![Screenshot 2021-07-18 153415](https://user-images.githubusercontent.com/1230712/126069138-9fb5ef22-e45c-444f-bc96-618a7978daa6.png)

Those LOG_ macros make it hard for the linter to detect. I'll see if I can override this in the standalone version so when compiling with pio it'll find them easier.
